### PR TITLE
Add support for dynamic feature modules

### LINKED ISSUE #55: No report for Dynamic feature module
### Current behavior

Cannot find any report for all dynamic feature module 

### Expected behavior

Report showing dynamic feature module size

### Steps to reproduce

Analyze project with DFM module

### Versions

1.1.0

### Anything else?

Hi, I am not sure why my DFM module not showing up on the report. It seems only showing base apk, is it expected or there some additional settings we need to add if want to analyze DFM module?

Thank you in advance

### PR DESCRIPTION
**Pull request by simonschiller**
### What has changed

- We now support [dynamic feature modules](https://developer.android.com/guide/playcore/feature-delivery) by having a dedicated section in the JSON and HTML report if modules are present in the app that's being analyzed.

### Why was it changed

- To be able to analyze DFMs as well as the main app. Note that DFMs are separated from the main app in both reports and not grouped together to make separate analysis easier.

### Notes
- This PR is mostly based on #81 and some discussions we had there. Huge thanks to @BaeBae33 for the initial effort!
- This turned out to be a pretty huge PR, reviewing by commit should make it easier to understand the changes.
- Maybe it's worth testing these changes on a more sophisticated app - like the Spotify app - before merging (cc @ViktorPetrovski @botronic).

### Screenshots
Capture

### Related issues

- Closes #55
- Closes #81

### CODE REVIEW

File: ruler-gradle-plugin/src/main/kotlin/com/spotify/ruler/plugin/dynamicfeature/DynamicFeatureHandler.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return extension.dynamicFeatures.map { feature ->
+            project.rootProject.project(feature).name
+        }
+    }
+
+    /** Groups all given [apks] based on the dynamic feature that they belong to. */
+    fun groupApks(dynamicFeatures: List<String>, apks: List<File>): Map<String, List<File>> {
+        val keys = (dynamicFeatures + BASE_FEATURE_NAME).sortedByDescending(String::length) // Sort to avoid ambiguity
+        return apks.groupBy { apk ->
+            keys.firstOrNull { apk.name.startsWith(it) } ?: throw IllegalArgumentException("Could not group $apk")
```
Comment (BaeBae33) [inline ruler-gradle-plugin/src/main/kotlin/com/spotify/ruler/plugin/dynamicfeature/DynamicFeatureHandler.kt:39]: When we have dfm named "base-dfm", `startsWith` works not property.
You can refer to [this code](https://github.com/spotify/ruler/compare/main...BaeBae33:ruler:dfm#diff-2079660ca1e76b08e0294243229ac7c48ba0dff7e54637629bd78c8275cb4603R65-R73) for getting each module's apks. It refers to [ApkSetUtils](https://github.com/google/bundletool/blob/96669880398d3df3fc8cfe12c82f3e94fa7ef36e/src/test/java/com/android/tools/build/bundletool/testing/ApkSetUtils.java#L48). I think this way is stable.

Comment (simonschiller) [reply ruler-gradle-plugin/src/main/kotlin/com/spotify/ruler/plugin/dynamicfeature/DynamicFeatureHandler.kt]: You're right, that's a good idea. Change it!

**Comment by botronic** [comment]
I will take a look this week!

**Review by botronic — APPROVED** [review]
Unfortunately, I was not able to test this on the Spotify app, so I will work on fixing the underlying issue there but this looks really good and it should be good to merge. Thanks @BaeBae33 and @simonschiller for making this happen! 🎉

**Pull request merged**